rebalance/rebalance.py

```diff
@@ -620,16 +633,68 @@ def maybe_rebalance_pairs(ch1, ch2, failed_channels: list):
     return result
 
 
+def rebalance_pair_picker(threadid, channel_pairs: list, failed_channels: list):
+    plugin.threadids[threading.get_ident()] = threadid
+    result = {"success": False, "fee_spent": Millisatoshi(0)}
+    idle_count = 0
+    while not rebalance_stopping():
+        idle_count += 1
+        for pair in channel_pairs:
+            if rebalance_stopping():
+                return result
+            ch1 = pair[0]
+            ch2 = pair[1]
+            processed = pair[2]
+            if processed:
+                continue
+            if ch1["lock"].acquire(blocking=False):
+                if ch2["lock"].acquire(blocking=False):
+                    idle_count = 0
+                    pair[2] = True
+                    result = maybe_rebalance_pairs(ch1, ch2, failed_channels)
+                    ch2["lock"].release()
+                ch1["lock"].release()
+            if result["success"]:
+                plugin.log(f"Thread{get_thread_id_str()} restarts rebalance threads after successful rebalance")
+                plugin.rebalance_stop_by_thread = True
```

Comment:
Question: Do I get it right, that the first successful rebalanced pair of any thread will stop the entire pool execution?

Reply:
Yep, we restart the pool. We need to get the channels again, as the balances changed, and we should continue with the new balances.

It's important to note this is not a complete restart, because `failed_channels` remains the same. So if we cannot rebalance a pair, we will not try again.

Reply:
Okay, how is it restarted then? I can only see that it sets `rebalance_stop_by_thread` to `True`, which will stop it.
Which code does the restarting?

Reply:
After all the threads stopped, the execution returns into `maybe_rebalance_once`, which last line is: `plugin.rebalance_stop_by_thread = False`
Then the `rebalanceall_thread` will call `maybe_rebalance_once` again, and `maybe_rebalance_once` gets fresh channel states in the first line.

It's almost the same as before this modification. `maybe_rebalance_once` was returning with a successful result and starting again with fresh channel states – before this commit too.

Reply:
Ahhhh, thanks :D